File: hash_table/0136_single_number.js
/*
Given a non-empty array of integers nums, every element appears twice except for 
one. Find that single one.

You must implement a solution with a linear runtime complexity and use only 
constant extra space.
*/

// https://leetcode.com/problems/single-number/

// hash table
// time complexity O(n)
// space complexity O(n)
var singleNumber = function (nums) {
  let dict = new Map();
  for (let i = 0; i < nums.length; i++) {
    if (dict.has(nums[i])) {
      dict.delete(nums[i]);
    } else {
      dict.set(nums[i], 1);
    }
  }
  return dict.keys().next().value;
};

// sorting
// time complexity O(nlogn)
// space complexity O(1)
var singleNumber = function (nums) {
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] !== nums[i + 1] && nums[i] !== nums[i - 1]) return nums[i];
  }
};

File: general_algorithm/bit_operation/0136_single_number.py
# Given a non-empty array of integers nums, every element appears twice except for 
# one. Find that single one.

# You must implement a solution with a linear runtime complexity and use only 
# constant extra space.

# 1 <= nums.length <= 3 * 10^4
# -3 * 10^4 <= nums[i] <= 3 * 10^4
# Each element in the array appears twice except for one element which appears only once.

# https://leetcode.com/problems/single-number/
# 反面例子（XOR不适用）：https://leetcode.com/problems/divide-array-into-equal-pairs/

# x ^ x = 0, x ^ 0 = x
# x ^ y = y ^ x
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        ans = 0
        for num in nums:
            ans ^= num
        return ans